src/libcore/slice.rs

```diff
@@ -914,6 +951,27 @@ impl<'a, T> IterMut<'a, T> {
     pub fn into_slice(self) -> &'a mut [T] {
         make_mut_slice!(T => &'a mut [T]: self.ptr, self.end)
     }
+
+    // Helper function for IterMut::nth
+    fn iter_nth(&mut self, n: usize) -> Option<&'a mut T> {
+        match make_mut_slice!(T => &'a mut [T]: self.ptr, self.end).get_mut(n) {
+            Some(elem_ref) => if mem::size_of::<T>() == 0 {
```

Comment:
I ended up using a macro to work around the mut/const problem. I could have used custom traits but that would have been a lot of extra code for little functionality... Something like (untested):
```rust
trait Reference<T> {
    type Pointer;
    unsafe fn as_ptr(&self) -> Self::Pointer {
        self as Self::Pointer
    }
}

trait Pointer<T> {
    type Reference;
    unsafe fn as_ref(&self) -> Self::Reference {
        transmute(self)
    }
    unsafe fn as_slice_ref(&self) -> Self::Reference {
        if mem::size_of::<T>() == 0 {
            &mut *(1 as *mut _)
        } else {
            self.as_ref()
        }
    }
    unsafe fn slice_offset(n: isize) -> Self;
}

impl<T> Reference<T> for &T {
    type Pointer = *const T;
}

impl<T> Reference<T> for &mut T {
    type Pointer = *mut T;
}

impl<T> Pointer<T> for *const T {
    type Reference = &T;

    unsafe fn slice_offset(n: isize) -> Self {
        if mem::size_of::<T>() == 0 {
            transmute(self as usize + n)
        } else {
            self.offset(n)
        }
    }
}

impl<T> Pointer<T> for *mut T {
    type Reference = &mut T;

    unsafe fn slice_offset(n: isize) -> Self {
        if mem::size_of::<T>() == 0 {
            transmute(self as usize + n)
        } else {
            self.offset(n)
        }
    }
}
```